### PR TITLE
Editor External Media: improve perceived performance by rendering more loading placeholders

### PR DESCRIPTION
**Pull request by jakejones1984**
Previously enough loading placeholders were rendered to cover the end of the row. However, [users were reporting that the external media library loading was slow](https://github.com/Automattic/wp-calypso/issues/22681). After [discussion](https://github.com/Automattic/wp-calypso/issues/25581) design input suggested loading lots of placeholders than necessary would improve the perception of speed. This PR implements that feedback.

Steps to test:

**On `master`:**
1. Open Media modal within Editor
2. Choose "Google Photos"
3. Start scrolling down. 
4. Notice how it feels slow 

Now checkout this branch and repeat the process. Notice how much faster it feels now more placeholders are loaded at once.

# Todo

- [x] Fix failing tests

### CODE REVIEW

**Comment by ehg** [comment]
Could we add some screen capture GIFs here?

**Comment by jakejones1984** [comment]
Yes sorry @ehg, would have been good to illustrate the different visually. 

# Original
media_scrolling

As you can see sometimes you only get a few loading squares showing up. This stops you from scrolling which provides the feeling of being "stuck". Also it's easy to missing the loading squares as there aren't many of them.

# Revised
media_scrolling_improved

Now we added a number of loading squares as per design guidance from @drw158. This makes the entire scrolling experience feel a lot more fluid and less restrictive. Whilst the speed of loading is actually no faster it feels much quicker to the user.


# Questions

* @drw158 does this fit with your original proposal
* Are there too many squares now?

**Comment by drw158** [comment]
Much much better! I don't think there are too many loading squares.

Some things I notices that may or may not be a separate issue:

- Why is there a hard stop after just a couple of scrolls? The UI does not tell me why it's not loading any more even though I know there are more pictures.
- The thumbnails provided by Google already have a play icon, so there is conflict with our own video icon.

screen shot 2018-07-03 at 10 29 09 am

**Comment by jakejones1984** [comment]
Thanks @drw158. Let me have a look into those issues for you.

**Comment by johngodley** [comment]
@jakejones1984 yes it appears to be maxing out. I've previously tested this up to 1000 so it definitely used to work that far. This leads me to think that a change has since occurred in Calypso or the WordPress API that is somehow limiting this. It's also possible that Picasa itself has changed. Certainly worth exploring while you are digging into this part of the code.

**Comment by jakejones1984** [comment]
@johngodley I've done some testing and I can confirm Picasa _is_ maxing out at this request

`https://picasaweb.google.com/data/feed/api/user/default?alt=json&max-results=50&imgmax=d&thumbsize=200,320,800,320c&fields=openSearch:itemsPerPage,openSearch:startIndex,link,entry(gphoto:timestamp,gphoto:id,gphoto:albumid,gphoto:numphotos,gphoto:size,title,media:group(media:content,media:thumbnail,media:title),summary)&start-index=331&q=%27%27&kind=photo`

(note: I'm doing my testing using Postman, and to replicate you will need to oAuth yourself first) 

I get back a response from the Picasa API with absolutely no `entry` value at all. This means there are no photos being returned. This causes the WPCom API to (correctly) report that there are no more results. (**side note**: we need to update the UI to show a message when there are no more results - currently it just stops scrolling).

I suspect this issue is happening because we are defaulting to the `recent` folder path. 

My theory is that the Picassa API has a limit on what they allow you paginate through within the "recent" folder. Moreover, the concept of "recent" suggests that there will be an arbitary limit on the number of photos Google classify as "recent".

To test this I made a request for a specific folder/album which I know has over 2000 photos in it. I passed a `start-index` of `400` and sure enough the Picasa API returned results! Therefore I believe our bug could be either:

1. A change in/regression of the Picasa API since launch which they haven't communicated
2. A bug that was always there from the start but which was never noticed until now (unlikely).

My recommendation is that we talk to Google to try and find out:

1. If they have changed the API so there's a hard limit on paginating the "recent" folder.
2. If it's possible to request "all" photos in any way.

**Comment by johngodley** [comment]
> A change in/regression of the Picasa API since launch

I've also confirmed this behaviour myself, and it seems the new max is 300. It used to be 1000, so this seems to be Google thing.

> sidenote: we need to update the UI to show a message when there are no more results - currently it just stops scrolling
>  I've now added an clear notice (see screenshot below which shows the bottom half of the media list) that will appear when there are no more results to display

This is standard media library behaviour. The proposed changes here will affect the behaviour of the media library for everything, not just Google Photos. As such I think it should be done elsewhere.

Also to note that the notice will be shown even if you have a single photo, making it a bit overpowering.

> the concept of "recent" suggests that there will be an arbitary limit on the number of photos Google classify as "recent"

The 'recent' term is our concept, and Google is just returning photos in reverse date order so there shouldn't be a limit on what recent means.

> My recommendation is that we talk to Google

Existing restrictions/bugs in Picasa are going to remain - Google won't change anything outside of the new API.

My feeling is this adds more weight to using the new API. In the mean time we should reduce the current max limit check to 300, and then the existing `Use the search button to access more photos. You can search for dates, locations, and things.` message will appear - this is the current workaround.

**Comment by jakejones1984** [comment]
@johngodley Thank you for your feedback. My responses below:

> This is standard media library behaviour. The proposed changes here will affect the behaviour of the media library for everything, not just Google Photos. As such I think it should be done elsewhere.

Agreed. I've reverted this.

> The 'recent' term is our concept, and Google is just returning photos in reverse date order so there shouldn't be a limit on what recent means.

I understand. However, if you don't request a specific Album from the API it appears to default to a kind of "recent" folder. I cannot confirm this but the following evidence leads me towards that conclusion

1. Hard limit of 300 odd photos when not specifying a folder (current functionality)
2. No limit observed when specifying a folder (please see [this PR](https://github.com/Automattic/wp-calypso/pull/25763#issuecomment-403249565))

> Existing restrictions/bugs in Picasa are going to remain - Google won't change anything outside of the new API.

I wasn't clear here. I meant, we should see if they are willing to confirm whether there is a limit when not requesting a specific folder/album. I feel it would be good to _know_ for certain if possible.

> My feeling is this adds more weight to using the new API. 

Absolutely. For now however, please see my [Album PR](https://github.com/Automattic/wp-calypso/pull/25763#issuecomment-403249565) which provides a workaround to the hard limit.

> In the mean time we should reduce the current max limit check to 300, and then the existing Use the search button to access more photos. You can search for dates, locations, and things. message will appear - this is the current workaround.

I'm going to propose keeping this PR related to the performance. @drw158 raised a couple of issues which I should have addressed elsewhere. We seem to agree that goal of "improving perceived performance" has been met and the other issues raised are now being solved in other PRs or workstreams.

Nonetheless, if you're happy for me to flip the value from `1000` to `300` I'm happy to do this. Please do let me know.

File: client/my-sites/media-library/list-item-video.jsx [hunk 1]
```diff
@@ -23,11 +23,13 @@ export default class extends React.Component {
 		media: PropTypes.object,
 		maxImageWidth: PropTypes.number,
 		thumbnailType: PropTypes.string,
+		showIcon: PropTypes.bool,
```
Comment (ehg) [inline client/my-sites/media-library/list-item-video.jsx]: We should move this `showIcon` fix into its own PR.

**Comment by alisterscott** [comment]
Is this one close to being able to be merged?

**Comment by jakejones1984** [comment]
> Is this one close to being able to be merged?

@alisterscott Apologies for the delay here (been away). I believe it's nearly there. Just need @johngodley to confirm he's happy with `1000` => `300` change.

> Nonetheless, if you're happy for me to flip the value from 1000 to 300 I'm happy to do this. Please do let me know.

@johngodley Are you able to confirm you're happy with the above? Much appreciated

**Comment by johngodley** [comment]
> Are you able to confirm you're happy with the above? Much appreciated

Yep, let's move to 300.

**Comment by jakejones1984** [comment]
@johngodley The hard limit is resolved in [this PR](https://github.com/Automattic/wp-calypso/pull/26129) so we might need to merge both PRs at the same time. 

@alisterscott There's quite lot of PRs so apologies for any confusion.

**Comment by getdave** [comment]
This PR may be made redundant as part of the changes discussed at `p4hfux-4s0-p2#comment-6708`. Leaving it open for now to see if it's still valid post-changes.

**Pull request closed without merging**